crates/json/src/lib.rs

```diff
@@ -0,0 +1,32 @@
+use std::collections::BTreeMap;
+
+use drawbridge_hash::Hash;
+use serde::{Deserialize, Serialize};
+
+/// A directory
+///
+/// Note that this type is designed to be extensible. Therefore, the fields
+/// here represent the minimum required fields. Other fields may be present.
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Directory {
+    /// The directory entries map
+    pub entries: BTreeMap<String, Entry>,
+}
+
+impl Directory {
+    pub const TYPE: &'static str = "application/vnd.drawbridge.directory.v1+json";
+}
+
+/// A directory entry
+///
+/// Note that this type is designed to be extensible. Therefore, the fields
+/// here represent the minimum required fields. Other fields may be present.
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Entry {
+    /// The hash of this entry
+    pub hash: Hash,
```

Comment:
So `hash` is the hash of the metadata of the *node*, right (i.e. `Content-Type`, `Content-Length` and `ETag`)?

That means that additional fields like `from` are never part of the hash of the node itself, but always a part of hash of the containing node, correct?

So if we have a directory node `A` with a singular entry node `B` with name `b` constructed from `x`,

```
hash_node(B) = hash({contentLength:len(B.contents),contentType:B.mimeType,eTag:hash(B.contents)})
```

```
A_contents = {entries:{b:{from:x,hash:hash_node(B)}}}
```

```
hash_node(A) = hash({contentLength:len(A_contents),contentType:application/vnd.drawbridge.directory.v1+json,eTag:hash(A_contents)})
```

A node `B` can be addressed via `hash_node(B)`, which does not include the `from` or any other "custom" fields.

If `application/vnd.drawbridge.entry.v1+json` is an allowed node media type, then we now have 2 ways of storing the same data, e.g. a `from` field could have been specified as part of the entry record in directory node, in which case it is never part of the entry itself or anything related to it (e.g. it's hash). Another option would be to have an "opaque" entry record in the directory node and, rather, upload a `application/vnd.drawbridge.entry.v1+json` node, in which case `from` field would now be a part of the node hash itself and be part of the node contents.

I feel like there should only be one way to do this and if we do want to add a new media type and have the additional metadata be part of the node identifier (node's hash), then we should only store hashes of directory node entries and not any additional fields related to those nodes.

Apart from that, what is the benefit of explicit `entries` field? Do we expect to extend the top-level directory type?

So the simplest approach addressing the above would be to simply keep the directory format as-is (i.e. a flat JSON object mapping node names to contained node hashes)

Reply:
I actually wrote this commit twice. The first time I wrote it I threw it away precisely because of this same thought.

I then tried to implement a system where `Meta` would essentially contain:
* `Content-Hash`
* `Content-Type`
* `Content-Length`
* `Drawbridge-*`

This way you could upload custom `Drawbridge-*` headers. Then in the case of mirroring, you would add something like `Drawbridge-From: python:3.10.0`.

However, I realized this produced a very strange property. When you would mirror an item, it would modify the root of the thing you are mirroring. i.e. the mirrored root note would have a different hash.

Returning to an earlier conversation we had, there are two kinds of properties about any object:
* `inherent`
* `exherent`

The hash, mime and size of the file are all `inherent` properties. OTOH, `from` is an exherent property. The `from` property derives from how a particular node is used and therefore shouldn't be part of the node hash itself. Stated another way, I think both of these are valid extension models at the same time. We need both `inherent` and `exherent` properties. Both should have immutability. But supporting `exherent` properties means we need at least the `Entity` type.

If you like the header model I proposed above, then I propose we drop the top layer of the `Directory` object and implement both. So a directory would look like:

```json
{
  "application.py": { "hash": "sha256:..." },
  "python": {
    "hash": "sha256:...",
    "from": "python:3.10.0"
  }
}
```

... instead of:

```json
{
  "entries": {
    "application.py": { "hash": "sha256:..." },
    "python": {
      "hash": "sha256:...",
      "from": "python:3.10.0"
    }
  }
}
```

Then `exherent` properties are always recorded in the `Entry` and `inherent` properties are always recorded as `Drawbridge-*` headers. Thoughts?